File: rust/benchmarks/src/suites/performance.rs
use crate::adapters::{benchmark_executable, fixture_arguments, Adapter, Execution};
use crate::model::TimedSuite;
use crate::runner::{BenchmarkCase, BenchmarkRunner};
use crate::BenchmarkResult;
use futures::future::join_all;
use std::path::Path;

fn adapter_cases<F>(
    adapters: &[Adapter],
    executable: &Path,
    arguments: &[String],
    validate: F,
) -> Vec<BenchmarkCase>
where
    F: Fn(&Execution) -> bool + Clone + 'static,
{
    adapters
        .iter()
        .map(|adapter| {
            let adapter = *adapter;
            let executable = executable.to_path_buf();
            let arguments = arguments.to_vec();
            let validate = validate.clone();
            BenchmarkCase::new(adapter.name(), move || {
                let executable = executable.clone();
                let arguments = arguments.clone();
                let validate = validate.clone();
                async move {
                    let result = adapter
                        .run(executable, &arguments)
                        .await
                        .map_err(|error| error.to_string())?;
                    validate(&result)
                        .then_some(())
                        .ok_or_else(|| format!("unexpected process result: {result:?}"))
                }
            })
        })
        .collect()
}

fn concurrent_cases(adapters: &[Adapter], executable: &Path, jobs: usize) -> Vec<BenchmarkCase> {
    adapters
        .iter()
        .map(|adapter| {
            let adapter = *adapter;
            let executable = executable.to_path_buf();
            BenchmarkCase::new(adapter.name(), move || {
                let executable = executable.clone();
                async move {
                    let operations = (0..jobs).map(|index| {
                        let arguments = fixture_arguments("echo", &[index.to_string()]);
                        let executable = executable.clone();
                        async move { adapter.run(executable, &arguments).await }
                    });
                    let results = join_all(operations).await;
                    for (index, result) in results.into_iter().enumerate() {
                        let result = result.map_err(|error| error.to_string())?;
                        let expected = serde_json::to_vec(&vec![index.to_string()])
                            .map_err(|error| error.to_string())?;
                        if result.exit_code != 0 || result.stdout != expected {
                            return Err(format!("unexpected concurrent result: {result:?}"));
                        }
                    }
                    Ok(())
                }
            })
        })
        .collect()
}

pub async fn run(
    runner: &BenchmarkRunner,
    adapters: &[Adapter],
    smoke: bool,
) -> BenchmarkResult<TimedSuite> {
    let executable = benchmark_executable()?;
    let output_bytes = if smoke { 64 * 1_024 } else { 1_024 * 1_024 };
    let jobs = if smoke { 2 } else { 8 };
    let overrides = smoke.then_some((2, 1));
    let mut scenarios = Vec::new();

    scenarios.push(
        runner
            .compare(
                "Process spawn latency",
                adapter_cases(
                    adapters,
                    &executable,
                    &fixture_arguments("echo", &["benchmark".to_string()]),
                    |result| result.exit_code == 0 && result.stdout == br#"["benchmark"]"#,
                ),
                overrides,
            )
            .await?,
    );
    scenarios.push(
        runner
            .compare(
                format!("Buffered stdout throughput ({output_bytes} bytes)"),
                adapter_cases(
                    adapters,
                    &executable,
                    &fixture_arguments("emit", &[output_bytes.to_string()]),
                    move |result| result.exit_code == 0 && result.stdout.len() == output_bytes,
                ),
                overrides,
            )
            .await?,
    );
    scenarios.push(
        runner
            .compare(
                format!("Concurrent execution ({jobs} processes)"),
                concurrent_cases(adapters, &executable, jobs),
                overrides,
            )
            .await?,
    );
    scenarios.push(
        runner
            .compare(
                "Non-zero exit handling",
                adapter_cases(
                    adapters,
                    &executable,
                    &fixture_arguments("fail", &["17".to_string()]),
                    |result| {
                        result.exit_code == 17 && result.stderr == b"intentional benchmark failure"
                    },
                ),
                overrides,
            )
            .await?,
    );
    scenarios.push(
        runner
            .compare(
                format!("command-stream output modes ({output_bytes} bytes)"),
                output_mode_cases(&executable, output_bytes),
                overrides,
            )
            .await?,
    );
    scenarios.push(
        runner
            .compare(
                format!("command-stream pipeline throughput ({output_bytes} bytes)"),
                pipeline_cases(&executable, output_bytes),
                overrides,
            )
            .await?,
    );
    scenarios.push(
        runner
            .compare(
                "command-stream built-in vs system process",
                built_in_cases(&executable),
                overrides,
            )
            .await?,
    );

    Ok(TimedSuite {
        kind: "performance".to_string(),
        name: "Performance".to_string(),
        scenarios,
    })
}

fn output_mode_cases(executable: &Path, bytes: usize) -> Vec<BenchmarkCase> {
    let buffered_executable = executable.to_path_buf();
    let streamed_executable = executable.to_path_buf();
    vec![
        BenchmarkCase::new("buffered", move || {
            let executable = buffered_executable.clone();
            async move {
                let arguments = fixture_arguments("emit", &[bytes.to_string()]);
                let result = command_stream::StreamingRunner::from_argv(executable, arguments)
                    .collect()
                    .await
                    .map_err(|error| error.to_string())?;
                (result.code == 0 && result.stdout.len() == bytes)
                    .then_some(())
                    .ok_or_else(|| "buffered output was incomplete".to_string())
            }
        }),
        BenchmarkCase::new("streaming", move || {
            let executable = streamed_executable.clone();
            async move {
                let arguments = fixture_arguments("emit", &[bytes.to_string()]);
                let mut stream =
                    command_stream::StreamingRunner::from_argv(executable, arguments).stream();
                let mut received = 0;
                let mut exit_code = None;
                while let Some(chunk) = stream.next().await {
                    match chunk {
                        command_stream::OutputChunk::Stdout(value) => received += value.len(),
                        command_stream::OutputChunk::Exit(value) => exit_code = Some(value),
                        command_stream::OutputChunk::Stderr(_) => {}
                    }
                }
                (received == bytes && exit_code == Some(0))
                    .then_some(())
                    .ok_or_else(|| "streamed output was incomplete".to_string())
            }
        }),
    ]
}

fn pipeline_cases(executable: &Path, bytes: usize) -> Vec<BenchmarkCase> {
    let pipeline_executable = executable.to_path_buf();
    let manual_executable = executable.to_path_buf();
    vec![
        BenchmarkCase::new("Pipeline API", move || {
            let executable = pipeline_executable.clone();
            async move {
                let source = command_line(
                    &executable,
                    &fixture_arguments("emit", &[bytes.to_string()]),
                );
                let destination = command_line(&executable, &fixture_arguments("stdin-count", &[]));
                let result = command_stream::Pipeline::new()
                    .add(source)
                    .add(destination)
                    .mirror_output(false)
                    .run()
                    .await
                    .map_err(|error| error.to_string())?;
                (result.code == 0 && result.stdout == bytes.to_string())
                    .then_some(())
                    .ok_or_else(|| format!("unexpected pipeline result: {result:?}"))
            }
        }),
        BenchmarkCase::new("manual two-step", move || {
            let executable = manual_executable.clone();
            async move {
                let source = command_stream::StreamingRunner::from_argv(
                    executable.clone(),
                    fixture_arguments("emit", &[bytes.to_string()]),
                )
                .collect()
                .await
                .map_err(|error| error.to_string())?;
                let destination = command_stream::StreamingRunner::from_argv(
                    executable,
                    fixture_arguments("stdin-count", &[]),
                )
                .stdin(source.stdout)
                .collect()
                .await
                .map_err(|error| error.to_string())?;
                (destination.code == 0 && destination.stdout == bytes.to_string())
                    .then_some(())
                    .ok_or_else(|| format!("unexpected manual result: {destination:?}"))
            }
        }),
    ]
}

fn built_in_cases(executable: &Path) -> Vec<BenchmarkCase> {
    let executable = executable.to_path_buf();
    vec![
        BenchmarkCase::new("built-in echo", || async {
            let result = command_stream::commands::echo(command_stream::CommandContext::new(vec![
                "benchmark".to_string(),
            ]))
            .await;
            (result.code == 0 && result.stdout == "benchmark\n")
                .then_some(())
                .ok_or_else(|| "unexpected built-in echo output".to_string())
        }),
        BenchmarkCase::new("spawned workload", move || {
            let executable = executable.clone();
            async move {
                let result = command_stream::StreamingRunner::from_argv(
                    executable,
                    fixture_arguments("echo", &["benchmark".to_string()]),
                )
                .collect()
                .await
                .map_err(|error| error.to_string())?;
                (result.code == 0 && result.stdout == r#"["benchmark"]"#)
                    .then_some(())
                    .ok_or_else(|| "unexpected spawned echo output".to_string())
            }
        }),
    ]
}

fn command_line(executable: &Path, arguments: &[String]) -> String {
    std::iter::once(shell_quote(&executable.to_string_lossy()))
        .chain(arguments.iter().map(|argument| shell_quote(argument)))
        .collect::<Vec<_>>()
        .join(" ")
}

#[cfg(unix)]
fn shell_quote(value: &str) -> String {
    command_stream::quote(value)
}

#[cfg(windows)]
fn shell_quote(value: &str) -> String {
    format!("\"{}\"", value.replace('"', "\\\""))
}

File: js/benchmarks/suites/performance.mjs
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { ProcessRunner, exec } from '../../src/$.mjs';

const benchmarkDirectory = dirname(dirname(fileURLToPath(import.meta.url)));
const fixture = join(benchmarkDirectory, 'fixtures', 'workload.mjs');

const casesFor = (adapters, args, validate) =>
  Object.fromEntries(
    adapters.map((adapter) => [
      adapter.name,
      {
        run: () => adapter.run(process.execPath, [fixture, ...args]),
        validate,
      },
    ])
  );

function concurrentCases(adapters, jobs) {
  return Object.fromEntries(
    adapters.map((adapter) => [
      adapter.name,
      {
        run: () =>
          Promise.all(
            Array.from({ length: jobs }, (_, index) =>
              adapter.run(process.execPath, [fixture, 'echo', String(index)])
            )
          ),
        validate: (results) =>
          results.every(
            (result, index) =>
              result.exitCode === 0 &&
              result.stdout === JSON.stringify([String(index)])
          ),
      },
    ])
  );
}

function streamCommand(bytes) {
  return new ProcessRunner(
    {
      mode: 'exec',
      file: process.execPath,
      args: [fixture, 'emit', String(bytes)],
    },
    { capture: true, mirror: false, stdin: 'ignore' }
  );
}

async function consumeStream(bytes) {
  let received = 0;
  let exitCode = null;
  for await (const chunk of streamCommand(bytes).stream()) {
    if (chunk.type === 'stdout') {
      received += Buffer.byteLength(chunk.data);
    }
    if (chunk.type === 'exit') {
      exitCode = chunk.code;
    }
  }
  return { received, exitCode };
}

async function bufferedCommand(bytes) {
  const result = await exec(
    process.execPath,
    [fixture, 'emit', String(bytes)],
    {
      capture: true,
      mirror: false,
      stdin: 'ignore',
    }
  );
  return { received: Buffer.byteLength(result.stdout), exitCode: result.code };
}

function fixtureRunner(mode, value, options = {}) {
  return new ProcessRunner(
    {
      mode: 'exec',
      file: process.execPath,
      args: [fixture, mode, String(value)],
    },
    { capture: true, mirror: false, stdin: 'ignore', ...options }
  );
}

async function programmaticPipeline(bytes) {
  const result = await fixtureRunner('emit', bytes).pipe(
    fixtureRunner('stdin-count', '', { stdin: 'pipe' })
  );
  return { exitCode: result.code, received: result.stdout };
}

async function bufferedPipeline(bytes) {
  const source = await fixtureRunner('emit', bytes);
  const destination = await fixtureRunner('stdin-count', '', {
    stdin: source.stdout,
  });
  return { exitCode: destination.code, received: destination.stdout };
}

async function builtInEcho() {
  const result = await exec('echo', ['benchmark'], {
    capture: true,
    mirror: false,
    stdin: 'ignore',
  });
  return result.stdout.trim();
}

export async function runPerformanceSuite({ runner, adapters, smoke = false }) {
  const outputBytes = smoke ? 64 * 1024 : 1024 * 1024;
  const jobs = smoke ? 2 : 8;
  const scenarioOptions = smoke ? { iterations: 2, warmup: 1 } : {};
  const scenarios = [];

  scenarios.push(
    await runner.compare(
      'Process spawn latency',
      casesFor(
        adapters,
        ['echo', 'benchmark'],
        (result) => result.exitCode === 0 && result.stdout === '["benchmark"]'
      ),
      scenarioOptions
    )
  );
  scenarios.push(
    await runner.compare(
      `Buffered stdout throughput (${outputBytes} bytes)`,
      casesFor(
        adapters,
        ['emit', String(outputBytes)],
        (result) =>
          result.exitCode === 0 &&
          Buffer.byteLength(result.stdout) === outputBytes
      ),
      scenarioOptions
    )
  );
  scenarios.push(
    await runner.compare(
      `Concurrent execution (${jobs} processes)`,
      concurrentCases(adapters, jobs),
      scenarioOptions
    )
  );
  scenarios.push(
    await runner.compare(
      'Non-zero exit handling',
      casesFor(
        adapters,
        ['fail', '17'],
        (result) =>
          result.exitCode === 17 &&
          result.stderr === 'intentional benchmark failure'
      ),
      scenarioOptions
    )
  );
  scenarios.push(
    await runner.compare(
      `command-stream output modes (${outputBytes} bytes)`,
      {
        buffered: {
          run: () => bufferedCommand(outputBytes),
          validate: ({ received, exitCode }) =>
            received === outputBytes && exitCode === 0,
        },
        streaming: {
          run: () => consumeStream(outputBytes),
          validate: ({ received, exitCode }) =>
            received === outputBytes && exitCode === 0,
        },
      },
      scenarioOptions
    )
  );
  scenarios.push(
    await runner.compare(
      `command-stream pipeline throughput (${outputBytes} bytes)`,
      {
        'pipe() API': {
          run: () => programmaticPipeline(outputBytes),
          validate: ({ exitCode, received }) =>
            exitCode === 0 && received === String(outputBytes),
        },
        'manual two-step': {
          run: () => bufferedPipeline(outputBytes),
          validate: ({ exitCode, received }) =>
            exitCode === 0 && received === String(outputBytes),
        },
      },
      scenarioOptions
    )
  );
  scenarios.push(
    await runner.compare(
      'command-stream built-in vs system process',
      {
        'built-in echo': {
          run: builtInEcho,
          validate: (output) => output === 'benchmark',
        },
        'spawned workload': {
          run: async () => {
            const result = await exec(
              process.execPath,
              [fixture, 'echo', 'benchmark'],
              { capture: true, mirror: false, stdin: 'ignore' }
            );
            return result.stdout;
          },
          validate: (output) => output === '["benchmark"]',
        },
      },
      scenarioOptions
    )
  );

  return {
    kind: 'performance',
    name: 'Performance',
    scenarios,
  };
}
